### DISCUSSION TITLE
Optimizing DuckDB Insert Performance: Parallelism and Row Groups

### DISCUSSION THREAD

**Discussion by hemathkumar-bitcot** [duckdb/duckdb#13371] [Q&A]
We're experiencing slow insert performance in DuckDB and are looking for advice on optimization. 

Here's our current situation:

1. Insert process takes about 12 minutes
2. Our server has 8 CPU cores
3. We understand parallelism is linked to the number of row groups
     example: 
     groups of rows that are stored together at the storage level. A row group in duck-db format consists of max 122,880 row 

Questions:
1. How can we verify if multiple row groups are being created during our insert process?
2. Is there a way to confirm if each row group is being processed in parallel on the available cores?
3. Are there any configuration settings or best practices to ensure optimal use of our 8 CPU cores during inserts?
4. What strategies can we employ to improve insert performance, particularly in relation to row groups and parallelism?

Any insights or suggestions would be greatly appreciated. Thanks in advance for your help!

**Comment by szarnyasg** [comment]
Hi @hemathkumar-bitcot, which API (command line, Python, Java, etc.) do you use? A brief import guide is available here: https://duckdb.org/docs/guides/performance/import

**Reply by hemathkumar-bitcot** [reply]
@szarnyasg we are using the command line.

**Reply by szarnyasg** [reply]
Do you use single INSERT statements? If so [wrap them in calls to `BEGIN TRANSACTION` and `COMMIT` as suggested in the [`INSERT` page](https://duckdb.org/docs/data/insert).

**Reply by hemathkumar-bitcot** [reply]
@szarnyasg , we would appreciate your insights on the following items. Thank you in advance for your assistance!

How can we verify if multiple row groups are being created during our insert process?
Is there a way to confirm if each row group is being processed in parallel on the available cores?
Are there any configuration settings or best practices to ensure optimal use of our 8 CPU cores during inserts?

**Comment by ryanhamilton** [comment]
I was trying to find more info and landed on this thread which wasn't useful.
I've added what benchmarks I ran myself here: https://www.timestored.com/data/duckdb/insert-benchmark

**Reply by michaeleengel1023** [reply]
I wonder what your performance is if you either 1) attach a new sqlite db to a new schema,  insert the data into sqlite then import from sqlite schema into duckdb table or 2) Insert into sqlite db, then import into duckdb? I had to use option 2 with a csv file with a multi-character delimiter. Below are my times using a 7 column, 1M row file tried with individual  inserts then trying with sqlite as an intermediate staging db.  Take this for whatever you feel it's worth.

7 columns, 1M rows.
direct csv import: 2.063 seconds
Individual inserts took: 886.2324030399323 seconds
Insert to sqlite/Import to duckdb: 6.561312437057495 seconds.
